File: dify-agent-sandbox-runtime/internal/server/output.go
package server

import (
	"fmt"
	"os"
	"unicode/utf8"
)

// OutputWindow represents a UTF-8-safe slice of output.log.
type OutputWindow struct {
	Output    string `json:"output"`
	Offset    int    `json:"offset"`
	Truncated bool   `json:"truncated"`
}

// ReadOutputWindow reads a forward UTF-8-safe slice from an output file.
func ReadOutputWindow(path string, offset, limit int) (*OutputWindow, error) {
	info, err := os.Stat(path)
	if os.IsNotExist(err) {
		if offset == 0 {
			return &OutputWindow{Output: "", Offset: 0, Truncated: false}, nil
		}
		return nil, NewServerError(400, "invalid_offset", "offset exceeds current file size 0")
	}
	if err != nil {
		return nil, err
	}

	size := int(info.Size())
	if offset > size {
		return nil, NewServerError(400, "invalid_offset",
			fmt.Sprintf("offset %d exceeds current file size %d", offset, size))
	}
	if offset == size {
		return &OutputWindow{Output: "", Offset: offset, Truncated: false}, nil
	}

	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	// Read up to limit+4 bytes to handle UTF-8 boundary
	readSize := limit + 4
	if offset+readSize > size {
		readSize = size - offset
	}

	buf := make([]byte, readSize)
	_, err = f.ReadAt(buf, int64(offset))
	if err != nil {
		return nil, err
	}

	// Advance past any UTF-8 continuation bytes at the start
	startShift := advanceToUTF8Boundary(buf, 0)
	data := buf[startShift:]
	budget := limit - startShift
	if budget < 0 {
		budget = 0
	}

	// Find the longest valid UTF-8 prefix within budget
	consumed := validUTF8PrefixLen(data, budget)
	if consumed == 0 && len(data) > 0 {
		// At least consume one complete rune
		consumed = firstCompleteRuneLen(data)
	}

	outputBytes := data[:consumed]
	newOffset := offset + startShift + consumed
	truncated := newOffset < size

	return &OutputWindow{
		Output:    string(outputBytes),
		Offset:    newOffset,
		Truncated: truncated,
	}, nil
}

// TailOutputWindow reads a UTF-8-safe tail snapshot from an output file.
func TailOutputWindow(path string, limit int) (*OutputWindow, error) {
	info, err := os.Stat(path)
	if os.IsNotExist(err) {
		return &OutputWindow{Output: "", Offset: 0, Truncated: false}, nil
	}
	if err != nil {
		return nil, err
	}

	size := int(info.Size())
	if size == 0 {
		return &OutputWindow{Output: "", Offset: 0, Truncated: false}, nil
	}

	start := size - limit
	if start < 0 {
		start = 0
	}
	paddedStart := start - 4
	if paddedStart < 0 {
		paddedStart = 0
	}

	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	readLen := size - paddedStart
	buf := make([]byte, readLen)
	_, err = f.ReadAt(buf, int64(paddedStart))
	if err != nil {
		return nil, err
	}

	relativeStart := advanceToUTF8Boundary(buf, start-paddedStart)
	payload := buf[relativeStart:]
	consumed := validUTF8PrefixLen(payload, len(payload))
	outputBytes := payload[:consumed]

	return &OutputWindow{
		Output:    string(outputBytes),
		Offset:    paddedStart + relativeStart + consumed,
		Truncated: false,
	}, nil
}

// advanceToUTF8Boundary skips continuation bytes (10xxxxxx) at position start.
func advanceToUTF8Boundary(data []byte, start int) int {
	for start < len(data) && isUTF8Continuation(data[start]) {
		start++
	}
	return start
}

// isUTF8Continuation returns true if byte is a UTF-8 continuation byte.
func isUTF8Continuation(b byte) bool {
	return (b & 0xC0) == 0x80
}

// validUTF8PrefixLen returns the length of the longest valid UTF-8 prefix up to maxLen.
func validUTF8PrefixLen(data []byte, maxLen int) int {
	if maxLen > len(data) {
		maxLen = len(data)
	}
	end := maxLen
	for end > 0 {
		if utf8.Valid(data[:end]) {
			return end
		}
		end--
	}
	return 0
}

// firstCompleteRuneLen returns the byte length of the first complete rune in data.
func firstCompleteRuneLen(data []byte) int {
	for end := 1; end <= len(data) && end <= 4; end++ {
		if utf8.Valid(data[:end]) {
			return end
		}
	}
	return 0
}

File: dify-agent/src/shellctl/shared/output.py
"""UTF-8-safe output slicing helpers for shellctl."""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path


@dataclass(slots=True, frozen=True)
class OutputWindow:
    """UTF-8-safe slice of `output.log`."""

    output: str
    offset: int
    truncated: bool


def read_output_window(path: Path, *, offset: int, limit: int) -> OutputWindow:
    """Read a forward UTF-8-safe slice from `output.log`."""

    if offset < 0:
        raise ValueError(f"offset must be >= 0; got {offset}")
    if limit <= 0:
        raise ValueError(f"limit must be > 0; got {limit}")

    if not path.exists():
        if offset == 0:
            return OutputWindow(output="", offset=0, truncated=False)
        raise ValueError(f"offset {offset} exceeds current file size 0")

    size = path.stat().st_size
    if offset > size:
        raise ValueError(f"offset {offset} exceeds current file size {size}")
    if offset == size:
        return OutputWindow(output="", offset=offset, truncated=False)

    with path.open("rb") as handle:
        handle.seek(offset)
        raw = handle.read(limit + 4)

    start_shift = _advance_to_utf8_boundary(raw, 0)
    data = raw[start_shift:]
    budget = max(0, limit - start_shift)
    consumed = _valid_utf8_prefix_length(data[:budget])
    if consumed == 0 and data:
        consumed = _first_complete_utf8_char_length(data)
    output_bytes = data[:consumed]
    new_offset = offset + start_shift + consumed
    truncated = new_offset < size
    return OutputWindow(
        output=output_bytes.decode("utf-8", errors="strict"),
        offset=new_offset,
        truncated=truncated,
    )


def tail_output_window(path: Path, *, limit: int) -> OutputWindow:
    """Read a UTF-8-safe tail snapshot from `output.log`."""

    if limit <= 0:
        raise ValueError(f"limit must be > 0; got {limit}")
    if not path.exists():
        return OutputWindow(output="", offset=0, truncated=False)
    size = path.stat().st_size
    if size == 0:
        return OutputWindow(output="", offset=0, truncated=False)
    start = max(0, size - limit)
    padded_start = max(0, start - 4)
    with path.open("rb") as handle:
        handle.seek(padded_start)
        raw = handle.read(size - padded_start)
    relative_start = _advance_to_utf8_boundary(raw, start - padded_start)
    payload = raw[relative_start:]
    consumed = _valid_utf8_prefix_length(payload)
    output_bytes = payload[:consumed]
    return OutputWindow(
        output=output_bytes.decode("utf-8", errors="strict"),
        offset=padded_start + relative_start + consumed,
        truncated=False,
    )


def _advance_to_utf8_boundary(data: bytes, start: int) -> int:
    while start < len(data) and _is_utf8_continuation_byte(data[start]):
        start += 1
    return start


def _valid_utf8_prefix_length(data: bytes) -> int:
    end = len(data)
    while end >= 0:
        try:
            data[:end].decode("utf-8", errors="strict")
            return end
        except UnicodeDecodeError as exc:
            if exc.start < end - 4:
                end = exc.start
            else:
                end -= 1
    return 0


def _is_utf8_continuation_byte(value: int) -> bool:
    return (value & 0b1100_0000) == 0b1000_0000


def _first_complete_utf8_char_length(data: bytes) -> int:
    for end in range(1, len(data) + 1):
        try:
            decoded = data[:end].decode("utf-8", errors="strict")
        except UnicodeDecodeError:
            continue
        if len(decoded) == 1:
            return end
        if decoded:
            return len(decoded[0].encode("utf-8"))
    return 0


__all__ = ["OutputWindow", "read_output_window", "tail_output_window"]
